README.md
# parser-6502

[![Greenkeeper badge](https://badges.greenkeeper.io/emkay/parser-6502.svg)](https://greenkeeper.io/)
Parser for 6502 assembler

This is a parser written in javascript that will parse assembler written for the [6502 8-bit microprocessor](https://en.wikipedia.org/wiki/MOS_Technology_6502).

[![Build Status](https://travis-ci.org/emkay/parser-6502.svg?branch=master)](https://travis-ci.org/emkay/parser-6502)
[![Standard - JavaScript Style Guide](https://img.shields.io/badge/code%20style-standard-brightgreen.svg)](http://standardjs.com/)
[![Standard Version](https://img.shields.io/badge/release-standard%20version-brightgreen.svg)](https://github.com/conventional-changelog/standard-version)

## Usage

```javascript
const fs = require('fs')
const parser = require('parser-6502')
const input = fs.readFileSync('./asm.s', 'utf-8')
const result = parser(input)

console.log(result)
```

## API

### result = parser(input)

```javascript
[
  {
    "directive": ".db",
    "args": [
      {
        "binary": "00000001"
      }
    ]
  }
]
```

## Installation

`npm i parser-6502`

## License

MIT


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.0.3](https://github.com/emkay/parser-6502/compare/v1.0.1...v1.0.3) (2019-06-14)


### Bug Fixes

* **audit:** audit fix update + travis node version ([ee2d2dc](https://github.com/emkay/parser-6502/commit/ee2d2dc))



<a name="1.0.2"></a>
## [1.0.2](https://github.com/emkay/parser-6502/compare/v1.0.1...v1.0.2) (2016-10-13)



<a name="1.0.1"></a>
## 1.0.1 (2016-10-13)


test/label.js
const tap = require('tap')
const mona = require('mona')
const labelParser = require('../parsers/label')

tap.test('should parse a label', (t) => {
  t.plan(1)
  t.deepEqual(mona.parse(labelParser(), 'SOMETHING:\n'), {
    label: 'SOMETHING'
  })
})


test/assembler.js
const tap = require('tap')
const parser = require('..')

tap.test('should parse the basics', (t) => {
  t.plan(1)
  const input = '.org $C000\nRESET:\n'
  t.deepEqual(parser(input), [
    {
      directive: '.org',
      args: [
        {
          'address': 'C000'
        }
      ]
    },
    {
      label: 'RESET'
    }
  ])
})


test/directive.js
const tap = require('tap')
const mona = require('mona')
const directiveParser = require('../parsers/directive')

tap.test('will parse a directive', (t) => {
  t.plan(1)
  t.deepEqual(mona.parse(directiveParser(), '.inesprg 1\n'), {
    args: [
      {
        'digit': '1'
      }
    ],
    directive: '.inesprg'
  })
})

tap.test('will parse a directive with direct address', (t) => {
  t.plan(1)
  t.deepEqual(mona.parse(directiveParser(), '.inesprg $0000\n'), {
    args: [
      {
        'address': '0000'
      }
    ],
    directive: '.inesprg'
  })
})

tap.test('will parse a directive with hex arg', (t) => {
  t.plan(1)
  t.deepEqual(mona.parse(directiveParser(), '.inesprg #$FE\n'), {
    args: [
      {
        'hex': 'FE'
      }
    ],
    directive: '.inesprg'
  })
})

tap.test('will parse a directive with binary arg', (t) => {
  t.plan(2)
  t.deepEqual(mona.parse(directiveParser(), '.db %00010001\n'), {
    args: [
      {
        'binary': '00010001'
      }
    ],
    directive: '.db'
  })

  t.deepEqual(mona.parse(directiveParser(), '.db #%00010001\n'), {
    args: [
      {
        'binary': '00010001'
      }
    ],
    directive: '.db'
  })
})

tap.test('will parse a directive with multiple args', (t) => {
  t.plan(1)
  t.deepEqual(mona.parse(directiveParser(), '.db %00010001,%00010001,%00010001\n'), {
    args: [
      {
        'binary': '00010001'
      },
      {
        'binary': '00010001'
      },
      {
        'binary': '00010001'
      }
    ],
    directive: '.db'
  })
})

tap.test('will parse a directive with multiple args with spaces between them', (t) => {
  t.plan(1)
  t.deepEqual(mona.parse(directiveParser(), '.db %00010001, %00010001, %00010001\n'), {
    args: [
      {
        'binary': '00010001'
      },
      {
        'binary': '00010001'
      },
      {
        'binary': '00010001'
      }
    ],
    directive: '.db'
  })
})

tap.test('will parse a directive with string arg', (t) => {
  t.plan(1)
  t.deepEqual(mona.parse(directiveParser(), '.incbin "mario.chr"\n'), {
    args: [
      {
        'string': 'mario.chr'
      }
    ],
    directive: '.incbin'
  })
})


parsers/parameters.js
const mona = require('mona')

function quotedChar () {
  return mona.or(
    mona.noneOf('"'),
    mona.and(
      mona.string('""'),
      mona.value('"')))
}

function digit () {
  return mona.text(mona.digit(), {min: 1})
}

function string () {
  return mona.between(
    mona.string('"'),
    mona.string('"'),
    mona.text(quotedChar(), {min: 1})
  )
}

function hex () {
  return mona.and(
    mona.string('#$'),
    mona.text(mona.digit(16), {min: 1})
  )
}

function binary () {
  return mona.and(
    mona.maybe(mona.string('#')),
    mona.string('%'),
    mona.text(mona.digit(2), {min: 1})
  )
}

function address () {
  return mona.and(
    mona.string('$'),
    mona.text(mona.alphanum(), {min: 1})
  )
}

// this is the fallthrough parser
function alphanum () {
  return mona.and(
    mona.not(digit()),
    mona.not(address()),
    mona.not(binary()),
    mona.not(hex()),
    mona.not(string()),
    mona.text(
      mona.alphanum(),
      {min: 1}
    )
  )
}

function parameter () {
  return mona.or(
    mona.tag(address(), 'address'),
    mona.tag(binary(), 'binary'),
    mona.tag(hex(), 'hex'),
    mona.tag(string(), 'string'),
    mona.tag(digit(), 'digit'),
    mona.tag(alphanum(), 'alphanum')
  )
}

function parameters () {
  return mona.split(
    parameter(),
    mona.trim(mona.string(','))
  )
}

module.exports = {
  parameter,
  parameters
}


parsers/instruction.js
const mona = require('mona')
const parameters = require('./parameters').parameters

const instructions = [
  'adc', 'and', 'asl',
  'bcc', 'bcs', 'beq', 'bit', 'bmi', 'bne', 'bpl', 'brk', 'bvc', 'bvs',
  'clc', 'cld', 'cli', 'clv', 'cmp', 'cpx', 'cpy',
  'dec', 'dex', 'dey',
  'eor',
  'inc', 'inx', 'iny',
  'jmp', 'jsr',
  'lda', 'ldx', 'ldy', 'lsr',
  'nop',
  'ora',
  'pha', 'php', 'pla', 'plp',
  'rol', 'ror', 'rti', 'rts',
  'sbc', 'sec', 'sed', 'sei', 'sta', 'stx', 'sty',
  'tax', 'tay', 'tsx', 'txa', 'txs', 'tya'
]

function instructionName () {
  return mona.oneOf(instructions)
}

function instruction () {
  return mona.sequence((s) => {
    const i = s(
      mona.followedBy(
        instructionName(),
        mona.maybe(mona.eol())
      )
    )

    const args = s(
      mona.maybe(
        mona.and(
          mona.spaces(),
          mona.followedBy(
            parameters(),
            mona.maybe(mona.eol())
          )
        )
      )
    )
    return mona.value({
      instruction: i,
      args: args
    })
  })
}

module.exports = instruction


example/example.s
.inesprg 1
.ineschr 1
.inesmap 0
.inesmir 1

.bank 0
.org $C000
RESET:
sei
cld
ldx #$40
stx $4017
ldx #$FF
txs
inx
stx $2000
stx $2001
stx $4010
vblankwait1:
bit $2002
bpl vblankwait1
clrmem:
lda #$00
sta $0000, x
sta $0100, x
sta $0200, x
sta $0400, x
sta $0500, x
sta $0600, x
sta $0700, x
lda #$FE
sta $0300, x
inx
bne clrmem
vblankwait2:
bit $2002
bpl vblankwait2

lda $2002
lda #$3F
sta $2006
lda #$00
sta $2006
ldx #$00
LoadPalettesLoop:
lda PaletteData, x
sta $2007
inx
cpx #$20
bne LoadPalettesLoop
LoadSprites:
ldx #$00
LoadSpritesLoop:
lda sprites, x
sta $0200, x
inx
cpx #$10
bne LoadSpritesLoop

LoadBackgrounds:
lda $2002
lda #$20
sta $2006
lda #$00
sta $2006
ldx #$00
LoadBackground1:
lda background1, x
sta $2007
inx
cpx #$00
bne LoadBackground1
LoadBackground2:
lda background2, x
sta $2007
inx
cpx #$00
bne LoadBackground2
LoadBackground3:
lda background3, x
sta $2007
inx
cpx #$00
bne LoadBackground3
LoadBackground4:
lda background4, x
sta $2007
inx
cpx #$00
bne LoadBackground4
LoadAttribute:
lda $2002
lda #$23
sta $2006
lda #$C0
sta $2006
ldx #$00
LoadAttributeLoop:
lda attribute, x
sta $2007
inx
cpx #$40
bne LoadAttributeLoop

lda #%10010000
sta $2000
lda #%00011110
sta $2001
Forever:
jmp Forever
NMI:

lda #$00
sta $2003
lda #$02
sta $4014

lda #$01
sta $4016
lda #$00
sta $4016
ReadA:
lda $4016
and #%00000001
beq ReadADone
ReadADone:
ReadB:
lda $4016
and #%00000001
beq ReadBDone
ReadBDone:
ReadSel:
lda $4016
and #%00000001
beq ReadSelDone
ReadSelDone:
ReadStart:
lda $4016
and #%00000001
beq ReadStartDone
ReadStartDone:
ReadUp:
lda $4016
and #%00000001
beq ReadUpDone
lda $0200
sec
sbc #$01
sta $0200
lda $0204
sec
sbc #$01
sta $0204
lda $0208
sec
sbc #$01
sta $0208
lda $020C
sec
sbc #$01
sta $020C
ReadUpDone:
ReadDown:
lda $4016
and #%00000001
beq ReadDownDone
lda $0200
clc
adc #$01
sta $0200
lda $0204
clc
adc #$01
sta $0204
lda $0208
clc
adc #$01
sta $0208
lda $020C
clc
adc #$01
sta $020C
ReadDownDone:
ReadLeft:
lda $4016
and #%00000001
beq ReadLeftDone
lda $0203
sec
sbc #$01
sta $0203
lda $0207
sec
sbc #$01
sta $0207
lda $020B
sec
sbc #$01
sta $020B
lda $020F
sec
sbc #$01
sta $020F
ReadLeftDone:
ReadRight:
lda $4016
and #%00000001
beq ReadRightDone
lda $0203
clc
adc #$01
sta $0203
lda $0207
clc
adc #$01
sta $0207
lda $020B
clc
adc #$01
sta $020B
lda $020F
clc
adc #$01
sta $020F
ReadRightDone:
lda #%10010000
sta $2000
lda #%00011110
sta $2001
lda #$00
sta $2005
sta $2005

rti
.bank 1
.org $E000

PaletteData:
.db $22,$29,$1A,$0F,$22,$36,$17,$0F,$22,$30,$21,$0F,$22,$27,$17,$0F
.db $0F,$16,$27,$18,$22,$02,$38,$3C,$22,$1C,$15,$14,$22,$02,$38,$3C

sprites:
.db $80, 112, $0, $80
.db $80, 113, $0, $88
.db $88, 114, $0, $80
.db $88, 115, $0, $88

background1:
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$45,$45,$24,$24,$45,$45,$45,$45,$45,$45,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$53,$54,$24,$24
.db $24,$24,$24,$24,$47,$47,$24,$24,$47,$47,$47,$47,$47,$47,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$55,$56,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
background2:
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$45,$45,$24,$24,$45,$45,$45,$45,$45,$45,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$53,$54,$24,$24
.db $24,$24,$24,$24,$47,$47,$24,$24,$47,$47,$47,$47,$47,$47,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$55,$56,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
background3:
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$45,$45,$24,$24,$45,$45,$45,$45,$45,$45,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$53,$54,$24,$24
.db $24,$24,$24,$24,$47,$47,$24,$24,$47,$47,$47,$47,$47,$47,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$55,$56,$24,$24
.db 17,14,21,21,24,36,32,24,27,21,13,36,36,36,36,36
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
background4:
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$45,$45,$24,$24,$45,$45,$45,$45,$45,$45,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$53,$54,$24,$24
.db $24,$24,$24,$24,$47,$47,$24,$24,$47,$47,$47,$47,$47,$47,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$55,$56,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
.db $24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24,$24
attribute:
.db %00010001, %00010001, %01010101, %00010001, %00010001, %00010001, %00010001, %01110111
.db %00010001, %00010001, %01010101, %00010001, %00010001, %00010001, %00010001, %01110111
.db %00010001, %00010001, %01010101, %00010001, %00010001, %00010001, %00010001, %01110111
.db %00010001, %00010001, %01010101, %00010001, %00010001, %00010001, %00010001, %01110111
.db %00010001, %00010001, %01010101, %00010001, %00010001, %00010001, %00010001, %01110111
.db %00010001, %00010001, %01010101, %00010001, %00010001, %00010001, %00010001, %01110111
.db %00010001, %00010001, %01010101, %00010001, %00010001, %00010001, %00010001, %01110111
.db %00010001, %00010001, %01010101, %00010001, %00010001, %00010001, %00010001, %01110111

.org $FFFA

.dw NMI
.dw RESET
.dw 0
.bank 2
.org $0000
.incbin "mario.chr"
